### PR TITLE
Switch to PDO

### PR DESCRIPTION
**Pull request by Apogate**
make saguaro use PDO & prepared statements. This should make things a lot safer overall as there's
no longer a need to escape strings, and any SQL queries without
hardcoded input are prepared now.

This is all through our own next to useless wrapper that I made which
definitely will need to be revisited as this is like the switch to
classes all over again for me.

This does NOT touch anything at all in the admin suite as its a fucking shitshow as always and will probably get rewritten soonish.

Still to do:

 - [ ] Rename `$mysql` (currently discussing)
 - [ ] Fix dangerous uses of `list()` and `extract()`, *highest priority to MySQL reads/writes*
   - I tried killing these wherever I saw them, there's still a few left probably
 - [ ] Start the switch over to ~~MySQLi~~ PDO

### CODE REVIEW

**Comment by Apogate** [comment]
Addressing comments in #254 

>I'm not sure if we can dynamically generate prepared statements, and we're handling a "lot" of data to begin with. Worst case scenario we'll have to manually update prepare statements as needed instead of it "just working" at all times. Beyond that, we might need to different setups to handle posts and their files.

Regist was able to generate dynamic prepared statements with some [minor tweaks](https://github.com/saguaroib/saguaro/blob/d9e4b6b753d0a237543d58c084461efa1775b2fe/_core/regist/regist.php#L125)

Regist's `dynamicBuild()` now returns 2 strings:
- `columns` ==> the SQL column names
- `values` ==> the prepared placeholder (ex `$no` would be `:no`)

and 1 array, the actual placeholder/value used to bind placeholders to values.
ex: `array(":no" => $no, ":name" => $name)`

**Comment by RePod** [comment]
Question mark

File: _core/admin/bans.php [hunk 1]
```diff
@@ -20,11 +20,10 @@
 class Banish {
     
     //If user is banned, return true.
-    function isBanned($ip, $redirect = 0) {
-        global $mysql;
-        
-        $exists   = $mysql->num_rows("SELECT * FROM " . SQLBANLOG . " WHERE host='$ip' AND (board='" . BOARD_DIR . "' or global=1)");
-        return ($exists > 0) ? true : false;
+    function isBanned($legacy = null) {
```
Comment (RePod) [inline _core/admin/bans.php:23]: Is `$legacy` even used? If not, why is here?

File: _core/delete/delete.php [hunk 2]
```diff
@@ -69,16 +69,16 @@ private function delete_post($no, $pwd, $imgonly = 0, $automatic = 0, $children
             //$mysql->query("INSERT INTO " . SQLDELLOG . " (admin, type, action, time, board, postno) VALUES ('{$auser}', '2', 'Deleted post #{$no}', '" . time() . "', '" . BOARD_DIR . "', '{$no}')");
         }
         
-        $conditions = ($row['resto'] == 0 && $children && !$imgonly) ? "(no='{$no}' OR resto='{$no}')" : "no='{$no}'"; 
-        $result = $mysql->query("SELECT * FROM " . SQLLOG . " WHERE {$conditions} AND board='" . BOARD_DIR . "'");
+        $conditions = ($row['resto'] == 0 && $children && !$imgonly) ? "(no=:no OR resto=:no)" : "(no=:no)";
```
Comment (RePod) [inline _core/delete/delete.php:72]: I'm not sure, just seems longer than it needs to be. We're checking three conditions just to enable one.

File: _core/delete/delete.php [hunk 3]
```diff
@@ -136,34 +136,34 @@ public function prune_old() { //This does the old thread pruning
...
-                $result      = $mysql->query("SELECT no FROM " . SQLLOG . " WHERE sticky=0 AND resto=0 ORDER BY $exp_order ASC");
-                $threadcount = $mysql->num_rows($result);
-                while ($row = $mysql->fetch_array($result) and $threadcount > $maxthreads) {
-                    $this->delete_post($row['no'], 'trim', 0, 1, 1, 0); // imgonly=0, automatic=1, children=1
-                    $threadcount--;
+                $result = $mysql->fetch_assoc("SELECT no FROM " . SQLLOG . " WHERE sticky=0 AND resto=0 ORDER BY $exp_order ASC");
+                $threadcount = $mysql->num_rows("SELECT no FROM " . SQLLOG . " WHERE sticky=0 AND resto=0 ORDER BY $exp_order ASC");
+                while ($threadcount > $maxthreads) {
+                    foreach ($result as $row) {
+                        $this->delete_post($row['no'], 'trim', 0, 1, 1, 0); // imgonly=0, automatic=1, children=1
```
Comment (RePod) [inline _core/delete/delete.php:143]: Note to consider having `delete_post()` accept an object with these as keys.

File: _core/general/resredir.php [hunk 4]
```diff
@@ -22,11 +17,11 @@
...
 }
 
-$redirect = DATA_SERVER . BOARD_DIR . "/" . RES_DIR . "/" . (($resto == 0) ? $no : $resto) . PHP_EXT . '#' . $no;
+$redirectect = DATA_SERVER . BOARD_DIR . "/" . RES_DIR . "/" . (($redirect['resto'] == 0) ? $redirect['no'] : $redirect['resto']) . PHP_EXT . '#' . $redirect['no'];
 
-echo "<META HTTP-EQUIV='refresh' content='0;URL=$redirect'>";
+echo "<META HTTP-EQUIV='refresh' content='0;URL=$redirectect'>";
 
-if ($resto == "0") {
+if ($redirect['resto'] == "0") {
```
Comment (RePod) [inline _core/general/resredir.php:24]: `== "0"`? The most recent comparison checked it as an integer.

Pay no attention to the previous version of this line.

File: _core/mysql/pdo.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	public function connect() {
+        $sqlhost = SQLHOST; $database = SQLDB;
+		$this->connection = new PDO("mysql:host=$sqlhost;dbname=$database", SQLUSER, SQLPASS);
+
+		/*if (!$this->connection->ping()) {
+			$this->handleError(S_SQLCONF,'', true);
+		}*/
+	}
+
+	public function selectDatabase($database) {
```
Comment (Apogate) [inline _core/mysql/pdo.php:17]: fix this

**Pull request merged**